applications/graph/MeshGraphNet/GNNComponents.py

```diff
@@ -0,0 +1,240 @@
+        edge_features (Layer) : Expected shape (Batch, Num_edges, self.in_dim_edge)
+        edge_indices (Layer): Expected shape (Batch, Num_edges)
+      Returns: 
+        (Layer): Expected shape (Batch, Num_nodes, self.in_dim_node) 
+    """
+    self.instance += 1
+  
+    edge_feature_sum = lbann.Scatter(edge_features, target_edge_indices,
+                                     name=f"{self.name}_{self.instance}_scatter",
+                                     dims=[self.num_nodes, self.in_dim_edge],
+                                     axis=0)
+
+    x = lbann.Concatenation([node_features, edge_feature_sum],
+                             axis=1,
+                             name=f"{self.name}_{self.instance}_concat_features")
+    x = self.node_mlp(x)
+
+    return lbann.Sum(node_features, x,
+                     name=f"{self.name}_{self.instance}_residual_sum")
+
+
+class GraphProcessor(Module):
+  """ Graph processor module 
+  """
+
+  def __init__(self,
+               num_nodes,
+               mp_iterations=15,
+               in_dim_node=128, in_dim_edge=128,
+               hidden_dim_node=128, hidden_dim_edge=128,
+               hidden_layers_node=2, hidden_layers_edge=2,
+               norm_type=lbann.LayerNorm):
+    super().__init__()
+
+    self.blocks = []
+
+    for _ in range(mp_iterations):
+      node_processor = NodeProcessor(num_nodes=num_nodes,
+                                     in_dim_node=in_dim_node,
+                                     in_dim_edge=in_dim_edge,
+                                     hidden_dim=hidden_dim_node,
+                                     hidden_layers=hidden_layers_node,
+                                     norm_type=norm_type)
+
+      edge_processor = EdgeProcessor(in_dim_node=in_dim_node,
+                                     in_dim_edge=in_dim_edge,
+                                     hidden_dim=hidden_dim_edge,
+                                     hidden_layers=hidden_layers_edge,
+                                     norm_type=norm_type)
+
+      self.blocks.append((node_processor, edge_processor))
+
+  def forward(self,
+              node_features,
+              edge_features,
+              source_node_indices,
+              target_node_indices):
+    """
+      Args:
+        node_features (Layer) : Expected shape (Batch, num_nodes, self.in_dim_node) 
+        edge_features (Layer) : Expected shape (Batch, Num_edges, self.in_dim_edge)
+        source_node_indices (Layer) : Expected shape (Batch, num_edges) 
+        target_node_indices (Layer) : Expected shape (Batch, num_edges)
+      Returns: 
+        (Layer, Layer): Expected shape (Batch, Num_nodes, self.in_dim_node) and 
+          (Batch, num_edges, self.in_dim_edge)
+    """
+
+    for node_processor, edge_processor in self.blocks:
+      x = node_processor(node_features, edge_features, target_node_indices)
+      e = edge_processor(node_features, edge_features,
```

Comment:
Shouldn't the node processor work on the result of the processed edges at each step? (see Eq. 1 of the paper)

Reply:
Yup, you are right.  `node_processor` and `edge_processor` should be switched. The updated edge features are used for updating node features.